include/material/mohr_coulomb.tcc

```diff
@@ -0,0 +1,571 @@
+    dp_dq = dp_drho * std::sqrt(2. / 3.);
+  }
+  // Compute softening part
+  double dphi_dpstrain = 0.;
+  double dc_dpstrain = 0.;
+  (*softening) = 0.;
+  if (softening_ && epds > epds_peak_ && epds < epds_residual_) {
+    // Compute dPhi/dPstrain
+    dphi_dpstrain = (phi_residual_ - phi_peak_) / (epds_residual_ - epds_peak_);
+    // Compute dc/dPstrain
+    dc_dpstrain =
+        (cohesion_residual_ - cohesion_peak_) / (epds_residual_ - epds_peak_);
+    // Compute dF/dPstrain
+    double df_dphi =
+        std::sqrt(1.5) * rho *
+            ((sin(phi) * sin(theta + M_PI / 3.) /
+              (std::sqrt(3.) * cos(phi) * cos(phi))) +
+             (cos(theta + M_PI / 3.) / (3. * cos(phi) * cos(phi)))) +
+        (mean_p / (cos(phi) * cos(phi)));
+    double df_dc = -1.;
+    (*softening) =
+        (-1.) * ((df_dphi * dphi_dpstrain) + (df_dc * dc_dpstrain)) * dp_dq;
+  }
+}
+
+//! Compute stress
+template <unsigned Tdim>
+Eigen::Matrix<double, 6, 1> mpm::MohrCoulomb<Tdim>::compute_stress(
+    const Vector6d& stress, const Vector6d& dstrain,
+    const ParticleBase<Tdim>* ptr, mpm::dense_map* state_vars) {
+  // Get equivalent plastic deviatoric strain
+  const double& epds = (*state_vars).at("epds");
+  // Update MC parameters using a linear softening rule
+  if (softening_ && (epds - epds_peak_) > 0. && (epds_residual_ - epds) > 0.) {
+    (*state_vars)["phi"] =
+        phi_residual_ + ((phi_peak_ - phi_residual_) * (epds - epds_residual_) /
+                         (epds_peak_ - epds_residual_));
+    (*state_vars)["psi"] =
+        psi_residual_ + ((psi_peak_ - psi_residual_) * (epds - epds_residual_) /
+                         (epds_peak_ - epds_residual_));
+    (*state_vars)["cohesion"] =
+        cohesion_residual_ +
+        ((cohesion_peak_ - cohesion_residual_) * (epds - epds_residual_) /
+         (epds_peak_ - epds_residual_));
+  }
+  if (softening_ && (epds - epds_residual_) >= 0.) {
+    (*state_vars)["phi"] = phi_residual_;
+    (*state_vars)["psi"] = psi_residual_;
+    (*state_vars)["cohesion"] = cohesion_residual_;
+  }
+  //-------------------------------------------------------------------------
+  // Elastic-predictor stage: compute the trial stress
+  Vector6d trial_stress = stress + (this->de_ * dstrain);
+  // Compute stress invariants based on trial stress
+  this->compute_stress_invariants(trial_stress, state_vars);
+  // Compute yield function based on the trial stress
+  Eigen::Matrix<double, 2, 1> yield_function_trial;
+  auto yield_type_trial =
+      this->compute_yield_state(&yield_function_trial, (*state_vars));
+  // Return the updated stress in elastic state
+  if (yield_type_trial == mpm::mohrcoulomb::FailureState::Elastic)
+    return trial_stress;
+  //-------------------------------------------------------------------------
+  // Plastic-corrector stage: correct the stress back to the yield surface
+  // Define tolerance of yield function
+  double Tolerance = 1E-1;
```

Comment:
Is it a typo?

Reply:
I think this at least has to be E-6 for iteration.